modules/navier_stokes/test/tests/finite_volume/physics/restart/2d_channel_scalar_turbulence_init.i

```diff
@@ -0,0 +1,278 @@
+    ymax = ${side_length}
+    nx = 100
+    ny = 20
+  []
+  [fmg_restart]
+    type = FileMeshGenerator
+    file = user_ics.e
+    use_for_exodus_restart = true
+  []
+[]
+
+[Physics]
+  [NavierStokes]
+    [Flow]
+      [all_flow]
+        compressibility = 'incompressible'
+
+        density = ${rho}
+        dynamic_viscosity = ${mu}
+
+        inlet_boundaries = 'left'
+        momentum_inlet_types = 'fixed-velocity'
+        momentum_inlet_functors = '${inlet_velocity} 0'
+
+        wall_boundaries = 'top bottom'
+        momentum_wall_types = 'noslip noslip'
+
+        outlet_boundaries = 'right'
+        momentum_outlet_types = 'fixed-pressure'
+        pressure_functors = '0'
+
+        mass_advection_interpolation = 'average'
+        momentum_advection_interpolation = 'average'
+      []
+    []
+    [FluidHeatTransfer]
+      [all_energy]
+        thermal_conductivity = ${k}
+        specific_heat = ${cp}
+
+        energy_inlet_types = 'fixed-temperature'
+        energy_inlet_functors = '1'
+
+        energy_wall_types = 'heatflux heatflux'
+        energy_wall_functors = '0 0'
+
+        energy_advection_interpolation = 'average'
+      []
+    []
+    [ScalarTransport]
+      [all_scalar]
+        passive_scalar_names = 'scalar1 scalar2'
+
+        passive_scalar_diffusivity = '${diff_scalar} ${diff_scalar}'
+        passive_scalar_source = '0.1 0'
+        passive_scalar_coupled_source = '1; 0'
+        passive_scalar_coupled_source_coeff = '0.1; 0'
+
+        passive_scalar_inlet_types = 'fixed-value fixed-value'
+        passive_scalar_inlet_functors = '1; 0.1'
+
+        passive_scalar_advection_interpolation = 'average'
+      []
+    []
+    [Turbulence]
+      [keps]
+        fluid_heat_transfer_physics = 'all_energy'
+        scalar_transport_physics = 'all_scalar'
+        turbulence_handling = 'k-epsilon'
+
+        tke_name = TKE
+        tked_name = TKED
+
+        # Fluid properties
+        Pr_t = ${Pr_t}
+        Sc_t = '0.7'
+
+        # Model parameters
+        C1_eps = ${C1_eps}
+        C2_eps = ${C2_eps}
+        C_mu = ${C_mu}
+        sigma_k = ${sigma_k}
+        sigma_eps = ${sigma_eps}
+
+        # Wall parameters
+        turbulence_walls = 'top bottom'
+        bulk_wall_treatment = ${bulk_wall_treatment}
+        wall_treatment_eps = ${wall_treatment_eps}
+        wall_treatment_T = ${wall_treatment_tem}
+
+        # Numerical parameters
+        mu_t_as_aux_variable = false
+        k_t_as_aux_variable = false
+      []
+    []
+  []
+[]
+
+[Executioner]
+  type = Steady
+  solve_type = 'NEWTON'
+  petsc_options_iname = '-pc_type -pc_factor_shift_type'
+  petsc_options_value = 'lu NONZERO'
+  nl_abs_tol = 1e-3
+
+  line_search = 'none'
+[]
+
+[Problem]
+  solve = false
+  # Currently necessary when using Newton's method, somewhere in the K-epsilon equations we do not preserve
+  # the sparsity pattern when no initial conditions for velocity are used
+  error_on_jacobian_nonzero_reallocation = false
```

Comment:
You get an error saying that the sparsity pattern is too small? It is a serious bug. That might be the reason why we cannot do split mesh either. We have an option to adjust the pattern, right? Maybe that would be a solution. I know fixing this is not the purpose of this PR, but I wouldn't like this to make it in like this. I can help debug this.

Reply:
These sparsity pattern issues got in in the first K-epsilon merge, I am slowly fixing them. I am almost done except for that last case with the 0 velocity initialization, which I don't think it s important to fix since one can just set non-zero velocities to get around.

Reply:
This is very important to fix. I'll take a look today. I really don't think we use AD information to define the sparsity pattern. Those should be taken care of by the relationship managers. I will have to see those before I believe that by default we allow this behavior.

Reply:
I'll also tag @lindsayad in case I am too slow finding the answer.

Reply:
Sorry, I didn't see your comment before. I am certain that the RMs are used for the sparsity pattern. I confirmed it with @roystgnr . I am checking where this can go wrong on the physics side. It is also suspicious that FVDiffusion needs the modifications too. It is the only one that requires `ghost_layers=2+` instead of 1. My suspicion is that something is off with the RM assignment at physics-based restart. I'll come back with actual details.

Reply:
I don't think so. First because the timing is similar to what the action was using. 

And then because the ghosting works fine for distributed cases. 

As I mentioned earlier, the new situation is that we end up calling these kernels with a 0 coefficient in the first iteration. Because if you don't initialize the turbulent viscosity that is what you get in your turbulent diffusion terms (momentum and scalar)

Reply:
Then in subsequent iterations we stop calling with a 0 coefficient.

Reply:
Okay we don't use any AD-related information for the sparsity pattern it is defined using the relationship managers alone. Based on this, I think what happens is: we allocate the matrix with a given sparsity pattern, then we don't populate some of the 0 entries and Petsc eliminates the unused entries. This is the default petsc behavior. Will be very hard to work around this besides the changes you added.

Reply:
There must be something special using AD though because we are still sending in 0s when we do these little multiplications by 0

Reply:
No, the zeros are eliminated after assembly:
https://petsc.org/main/manualpages/Mat/MatEliminateZeros/
This is called which gets rid of every offdiagonal zero. 
I am still looking into the details for the precise call stack to make sure we know what is happening.

Reply:
Never mind, was looking at the wrong call stack. In theory, petsc should shrink the matrix only if we don't put anything in it. That's why putting 0s explicitly works. Will get the call stack soon.

Reply:
Once we're done here, let's add some documentation in the AD page and the relationship manager page.

Reply:
It is actually here:
https://petsc.org/release/manualpages/Mat/MatAssemblyBegin/
And the end routine too.
Actually, Fande added an option to restore the original preallocation before the next assembly.
https://petsc.org/release/manualpages/Mat/MatResetPreallocation/
Maybe we can rely on this when we restart. Let me check if we can use it.

Reply:
this code is not related to restarts, but rather to not initializing mu_t

Reply:
But allowing nonzero jacobian entry allocations on the fly will have a huge burden on computation time and should be avoided if possible. I am checking if it is possible.

Reply:
kk well checkout kernels with velocity & k / epsilon.
It's a long task, there are tons of BCs

Reply:
`MatResetPreallocation` is a great option if the sparsity pattern determined by the relationship managers is always a superset of the true nonzeroes and is not overly dense. This is generally not the case for things like mechanical contact for which the mesh deformation leads to the true sparsity pattern changing over the course of the simulation in which case your initial matrix preallocation would have to be very dense.

Reply:
When does a Jacobian get assembled when `solve = false`? It would be good to know whether you are getting a new nonzero for the initial Jacobian assembly or a subsequent one. If the first, then there is an issue with the relationship managers. I know that your Navier-Stokes physics adds 3 ghosting layers (corresponding to the `INSFVMixingLengthReynoldsStress` kernel). And I also know that different turbulence inputs require 5-7 ghosting layers perusing the input files.

If you are getting a new nonzero during a Jacobian assembly later than the first, then we either need to reset the preallocation like @grmnptr brought up or we need to ensure we are inserting zeroes during our Jacobian assembly where zeroes may arise in later Jacobian evaluations, e.g. for upwind schemes if the wind changes direction

Reply:
Okay, how about this:
- I think the ghost layers are probably fine, I think the inputs that require more are the two-phase models.
- Instead of using the early return just use the interpolation below that. I think we ran into trouble because of this early return exception. That will give you the proper derivatives no matter what. You can fall back to average if harmonic is doomed to fail.

If it still errors without `error_on_jacobian_nonzero_reallocation = false` we have a serious problem. 
I hope we won't have to touch the preallocation part. 


Unrelated, but ghosting layer of 7 looks really fishy to me. Feels like an infinite loop and ghosting layer 5-7 was the first that passed because everything was ghosted at that point. Looking at the input we have 4x10 elements in the mesh so it is very likely that we ghost everything. So we might have a big issue for the two-phase models as well.

Reply:
> When does a Jacobian get assembled when solve = false?

It does not, I just made sure the input ran if a user wanted to drop the solve = false and run it. 

>  or a subsequent one.

then.

This is simply because of bad zero ICs. I already took care of what mattered here, having a constant sparsity pattern with initialization and with K-Epsilon. Let's remember that the many inputs of ours (iirc msfr) do not run with 0 initial conditions for velocity. So not initializing (to 1e-8 for example) is not viable anyway. 

I know I did not check every single BC in K-epsilon but that's it

>  And I also know that different turbulence inputs require 5-7 ghosting layers perusing the input files.

yeah we should look into it at some point. Is it for turbulence though? I see 3 in the WCNSFV Turbulence physics. I think the 2 phase mixtures are affected. I just checked 2 k-epsilon inputs and they dont have 5-7

> we need to ensure we are inserting zeroes during our Jacobian assembly where zeroes may arise in later Jacobian evaluations

this is what I do here, to go around the user not initializing mu_t. 
If I dont do this, and the user makes an input with kernels, then they will hit the error. With Physics we could enforce that they initialize if we wanted

Reply:
> Instead of using the early return just use the interpolation below that. I think we ran into trouble because of this early return exception. That will give you the proper derivatives no matter what. You can fall back to average if harmonic is doomed to fail.

this crashes with harmonic.
I don't see how
```
if (coeff. == 0)
  do_average_interpolation
else
  do_interpolation_user_requested
```
is any better than what I have now. 

> If it still errors without error_on_jacobian_nonzero_reallocation = false we have a serious problem.

requiring users to set their initial velocities to 1e-8 is not a serious problem. 
If you have an example where things are initialized properly and we get errors, then I agree

Reply:
The issue is that we have these min max limiters here and there (including the early return). And those don't carry the derivatives. They are just numbers. So we have big differences between existing derivatives depending on the raw value. And yes, there is a big difference in terms of logic. That way this sparsity pattern preservation issue does not get exposed in user-facing functions. I am really second guessing that we should allow k-epsilon with Newton.

Reply:
In my opinion, users should not be aware of the sparsity pattern and should not learn about it either. I think it would be bad design if they had to worry about it in an fvdiffusion just because of the limiters in k epsilon discarding derivatives left and right.

Reply:
Also the fvdiffusion qp function is in the moose worksop slide and I think showcasing this with this sparsity pattern hackery would be embarassing compared to the fekernel for example.

Reply:
But it is a user facing problem. If you are looking at this code, it's because you are most likely coding a diffusion kernel. And if you are doing so and are not mindful of the sparsity pattern, then you ll get an error. 

if we go hide the multiplication by 0 or a sneaky switch to average inside the MathFVUtils.h we leave the user open to those issues. 

>  I am really second guessing that we should allow k-epsilon with Newton.

It has been allowed since it was merged. Its sparsity pattern issues are essentially fixed and there's even an action for it now. Seems like not the time to rip it out.

Reply:
I'm totally ok with a disclaimer that we dont recommend using it and that we are working on something better.

Reply:
> In my opinion, users should not be aware of the sparsity pattern and should not learn about it either.

at some point they do though. Coding in FV they will inevitably hit the issue.

>  I think it would be bad design if they had to worry about it in an fvdiffusion just because of the limiters in k epsilon discarding derivatives left and right.

again not that. mu_t = 0 is the source here.

Reply:
Okay how about:
- we remove this from fvdiffusion so that at least that is presentable and does not do it for this kepsilon solution alone
- make sure that we can run without the `allocation=false` option. 
I have a few ideas, will try them tomorrow. Will push a commit or 2.

Reply:
but you're right that there is some limiting in the functor material route. So maybe that is affected too. 
but again, mu_t = 0 is the route problem, which comes from users not initializing tke

> Also the fvdiffusion qp function is in the moose worksop slide and I think showcasing this with this sparsity pattern hackery would be embarassing compared to the fekernel for example.

I guess that's a good reason. It s too early for the simple FV tutorial. 
I ll think of something.

Reply:
I think adding the interpolation with an average override if harmonic is in trouble will solve the fvdiffusion side. I can hunt down some derivatives tomorrow.

Reply:
you sure your time would not be better spent writing some linear kernels for this :D ?

Reply:
hunting down derivatives for the theoretical case of users yolo-ing the initial conditions would not be my go to

Reply:
I think the fix might be actially not that hard. We should probably just add some derivatives manually when we limit. But yeah is that a nicer solution than this, most likely not. Ah man.

Reply:
to solve mu_t = 0 (true 0), let s just run the auxkernel on initial and add a limiter to it
then we can take off this logic

Reply:
Oh wait, I thought the one with newton's method used a functormaterial for mut. I'll check the implementation, this whole thread could have been a misunderstanding.

Reply:
ah it is
```
        mu_t_as_aux_variable = false
```

Reply:
interesting. 
Maybe it was hitting a zero mu_t (instead of a 1e-8) before I moved the limiter from the denominator to the numerator

Reply:
Or maybe it is the limiter then. 
So off of the lack of initialization of the tke.

Reply:
But regular harmonic interpolation would have trouble with it too. The early return isn't triggered when the viscosity is limited

Reply:
I think in that case the issue is that we won't have the derivatives of k and epsilon within mu_t in the first assembly. 
I'll take a look. Was it a hard requirement to not initialize tke?

Reply:
In an ideal world users should not have to think about sparsity pattern. They have not had to do so for the first 16 years of MOOSE.

I'm not saying that it's not good for them to think about it. I'm just saying ideally they shouldn't *have* to

Reply:
Like there was a time users had to set non-zero velocity for INSAD (and maybe for FV) to avoid undefined derivatives due to taking the norm of the velocity. I thought that didn't look very nice in the input so I slowly fixed that

Reply:
the lid-driven energy case also works now with no initial conditions